oauth/api/router.py

```diff
@@ -0,0 +1,345 @@
+# pylint: disable=unused-argument
+
+from datetime import datetime
+from json import loads as load
+from logging import getLogger
+from typing import Any, Callable, Generator, List, Optional
+
+from fastapi import APIRouter, Body, Depends, Request, Security, status
+from fastapi.routing import APIRoute
+from fastapi.security import HTTPBasic, SecurityScopes
+from sqlalchemy.orm import Session
+
+import oauth.api.endpoints as endpoints
+import oauth.jwt as jwt
+from oauth.api.exceptions import (
+    AuthenticationException,
+    AuthorizationException,
+    ClientNotFoundException,
+    ExpiredTokenException,
+)
+from oauth.api.models import (
+    AccessToken,
+    ClientCreatedResponse,
+    OAuth2ClientCredentialsBearer,
+    OAuth2ClientCredentialsRequestForm,
+)
+from oauth.api.utils import extract_payload, is_token_expired, validate_scopes
+from oauth.database.models import ClientDetail
+from oauth.scopes import (
+    CLIENT_CREATE,
+    CLIENT_DELETE,
+    CLIENT_READ,
+    CLIENT_UPDATE,
+    SCOPE_DOCS,
+    SCOPE_READ,
+    SCOPES,
+)
+
+EIGHT_DAYS = 60 * 24 * 8  # Expressed in minutes
+SORTED_SCOPES = SCOPES.sort()
+
+logger = getLogger(__name__)
+oauth2_scheme = OAuth2ClientCredentialsBearer(
+    tokenUrl=endpoints.TOKEN,
+    scopes=SCOPE_DOCS,
+)
+
+
+class OAuthRouter(APIRouter):
+    """
+    A FastAPI `APIRouter` that includes all endpoints necessary to
+    implement a complete OAuth scheme.
+    """
+
+    def __init__(
+        self,
+        app_encryption_key: str,
+        db: Callable[[], Generator[Session, None, None]],
+        *,
```

Comment:
i'll admit i've never seen this before, what does it do? i've seen *args and **kwargs

Reply:
The `*` indicates the end of the positional arguments. Every argument that comes afterwards can only be specified by keyword, as defined in [PEP-3102](https://www.python.org/dev/peps/pep-3102/).

Reply:
TIL!